File: typescript/src/transport/inProcess.ts
import type { AuditEvent } from '../schema/event.js';
import type { AuditCapability } from '../schema/capability.js';
import type { AuditHost } from '../host/auditHost.js';
import type { AttemptResponse, AuditTransport } from './transport.js';

// In-process transport: the tool calls the host directly. This is the only asset superseded
// when B2 swaps in an MCP-SDK-backed transport — and it survives as a fast test double.
// It models the wire contract faithfully (request/response for attempt, void for outcome,
// capability negotiation), so nothing above this line changes on the swap.
export class InProcessTransport implements AuditTransport {
  constructor(private readonly host: AuditHost) {}

  negotiate(): AuditCapability {
    return this.host.negotiate();
  }

  async sendAttempt(event: AuditEvent): Promise<AttemptResponse> {
    return this.host.handleAttempt(event);
  }

  async sendOutcome(event: AuditEvent): Promise<void> {
    this.host.handleOutcome(event);
  }
}

File: reference/python/src/auditable_mcp/in_process.py
"""In-process transport implementation.

Directly couples the tool to the host audit subsystem without network overhead.
Used primarily for testing and local demonstration.
"""

from auditable_mcp.capability import AuditCapability, NegotiationResult
from auditable_mcp.host import AuditHost
from auditable_mcp.transport import AttemptResponse


class InProcessTransport:
    """Direct in-process transport implementing the AuditTransport protocol."""

    def __init__(self, host: AuditHost) -> None:
        """Bind the transport to a host audit subsystem."""
        self._host = host

    def negotiate(self, offered: AuditCapability) -> NegotiationResult:
        """Forward the tool's offered capability to the host and return the negotiation result."""
        return self._host.negotiate(offered)

    def send_attempt(self, event: dict) -> AttemptResponse:
        """Forward audit/attempt to the host."""
        return self._host.handle_attempt(event)

    def send_outcome(self, event: dict) -> None:
        """Forward audit/outcome to the host."""
        self._host.handle_outcome(event)
